README
This file contains a tree and a explanation about the files of the project

trace
├── AUTHORS
├── ChangeLog
├── INSTALL
├── LICENSE
├── Makefile
├── NEWS
├── README
├── README.md
├── doc
│   └── img
│       └── trace.png
├── include
│   └── trace
│       └── trace.h
├── install.sh
├── trace.conf
├── man
│   └── libtrace.3
├── mk
├── mkall
├── mkclean
├── mkd
├── mkdall
├── mkdistclean
├── mkinstall
├── mktest
├── mkuninstall
├── remove_logs.sh
├── src
│   └── trace.c
├── tests
│   ├── colored.c
│   ├── debug.c
│   ├── hello_log.c
│   ├── module1.c
│   ├── module2.c
│   ├── trace.c
│   └── trace_level.c
└── uninstall.sh

8 directories, 32 files

AUTHORS.......: The name and email of project author
ChangeLog.....: Changelog of project
INSTALL.......: Installation guide
LICENSE.......: License of software
Makefile......: Makefile of the project
NEWS..........: File content the news about the project
README........: This file containing a explanation about the files of project
README.md.....: The README of github
trace.h.......: Header file of the library
install.sh....: Installation script
trace.conf....: A .conf example file to use the trace library in your C programs
libtrace.3....: A man page of the library and your functions
mk............: A make script
mkall.........: make all script
mkclean.......: make clean script
mkd...........: make debug script
mkdall........: make debug all script
mkdistclean...: make distclean script
mkinstall.....: make install script
mktest........: make test script
mkuninstall...: make uninstall script
remove_logs.sh: script to remove all .log files in current directory
trace.c.........: C source file of the library
uninstall.sh..: Uninstall script of the software



NEWS
# -*- org *-*

Copyright (C) 2023 Gustavo Bacagine <gustavo.bacagine@protonmail.com>

libtrace 2023-12-25
Changed the values of debug levels

libtrace 2023-11-30
Changed the name of the library and the source files

libtrace 2023-10-18
Changed the name of macros used in comparison of debug level

ligbrace 2023-10-17

Changed the name of macros, functions and variables

Add new macros:
 vTrace
 vTraceBegin
 vTraceEnd

libtrace 2023-09-22

Add the make scripts and the following scripts: 
 install.sh,
 uninstall.sh
 remove_logs.sh



tests/debug.c
/**
 * debug.c: Exemple of the DEBUG_DETAILS
 * 
 * Written by Gustavo Bacagine <gustavo.bacagine@protonmail.com>
 * 
 * Date: 2023-06-17
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include "trace.h"

#define UNUSED(X) (void) X

/**
 * Structure that 
 * represents a person
 */
typedef struct Person
{
  char szName[64];
  uint8_t uAge;
} Person, *PPerson;

static void vInitPerson(PPerson *ppstPerson)
{
  if(INFO_DETAILS)
  {
    vTraceInfo("start %s function", __func__);
    vTraceInfo("Setting the default values in members of struct Person");
  }
  
  memset((*ppstPerson)->szName, 0, sizeof((*ppstPerson)->uAge));
  (*ppstPerson)->uAge = 0;
  
  if(DEBUG_DETAILS)
  {
    vTraceDebug("memset((*ppstPerson)->szName, 0, sizeof((*ppstPerson)->uAge))");
    vTraceDebug("(*ppstPerson)->uAge = 0");
  }

  if(INFO_DETAILS)
  {
    vTraceInfo("end %s function", __func__);
  }
}
  
static void vCreatePerson(PPerson *ppstPerson, const char *szName, const uint8_t uAge)
{
  if(INFO_DETAILS)
  {
    vTraceInfo("start %s function", __func__);
    vTraceInfo("create a new person");
  }

  strcpy((*ppstPerson)->szName, szName);
  (*ppstPerson)->uAge = uAge;

  if(DEBUG_DETAILS)
  {
    vTraceDebug("Values of the member Person");
    vTraceDebug("(*ppstPerson)->szName = %s", (*ppstPerson)->szName);
    vTraceDebug("(*ppstPerson)->uAge = %d", (*ppstPerson)->uAge);
  }

  if(INFO_DETAILS)
  {
    vTraceInfo("Person created with succes");
    vTraceInfo("end %s function", __func__);
  }
}

static void vShowPerson(PPerson pstPerson)
{
  if(INFO_DETAILS)
  {
    vTraceInfo("start %s function", __func__);
    vTraceInfo("Show Person");
    vTraceInfo("Name: %s\n"
             "Age.: %d\n", pstPerson->szName,
                           pstPerson->uAge);
  }

  printf("Name: %s\n"
         "Age.: %d\n", pstPerson->szName,
                       pstPerson->uAge);

  if(INFO_DETAILS)
  {
    vTraceInfo("end %s function", __func__);
  }
}

static void vDestroyPerson(PPerson *ppstPerson)
{
  if(INFO_DETAILS)
  {
    vTraceInfo("start %s function", __func__);
  }
  
  memset((*ppstPerson)->szName, 0, sizeof((*ppstPerson)->uAge));
  (*ppstPerson)->uAge = 0;
  
  if(DEBUG_DETAILS)
  {
    vTraceDebug("memset((*ppstPerson)->szName, 0, sizeof((*ppstPerson)->uAge))");
    vTraceDebug("(*ppstPerson)->uAge = 0");
    vTraceDebug("Freeing Person structure pointer");
  }

  free((*ppstPerson));
  (*ppstPerson) = NULL;

  if(DEBUG_DETAILS)
  {
    vTraceDebug("free((*ppstPerson))");
    vTraceDebug("(*ppstPerson) = NULL");
    vTraceDebug("Person structure pointer free");
  }

  if(INFO_DETAILS)
  {
    vTraceInfo("end %s function", __func__);
  }
}

int main(int argc, char **argv)
{
  PPerson pstPerson = (PPerson) malloc(sizeof(Person));
  
  UNUSED(argc);
  UNUSED(argv);
  UNUSED(gbColoredLogLevel);
  UNUSED(gszLogFileName);
  UNUSED(gszConfFileName);
  UNUSED(kszLogLevelColorEnd);
  UNUSED(kszLogLevelColorInit);
  UNUSED(kszLogLevel);
  
  vSetConfFileName("trace.conf");
  
  vSetLogLevel(iGetLogLevel());

  if(giDebugLevel < 1)
  {
    fprintf(stderr, "Error, giDebugLevel return value: %d!\n", giDebugLevel);

    exit(EXIT_FAILURE);
  }
  
  vSetLogFileName("debug.log");
  
  /* Initial trace messages */
  if(INFO_DETAILS)
  {
    vTraceInfo("Start of function %s", __func__);
  }
  
  if(pstPerson == NULL)
  {
    if(DEBUG_DETAILS)
    {
      vTraceDebug("E: out memory in allocation of pstPerson!");
    }
  }

  vInitPerson(&pstPerson);
  
  vCreatePerson(&pstPerson, "Gustavo Bacagine", 23);
  
  vShowPerson(pstPerson);

  vDestroyPerson(&pstPerson);
  
  /* End trace messages */
  if(INFO_DETAILS)
  {
    vTraceInfo("End of function %s", __func__);
  }

  return 0;
}



tests/trace.c
/**
 * trace.c
 *
 * Written by Gustavo Bacagine <gustavo.bacagine@protonmail.com>
 * 
 * Description: File created to test the new macros:
 *  vTrace
 *  vTraceBegin
 *  vTraceEnd
 *
 * Date: 2023-10-17
 *
 */

#include <stdio.h>
#include "log.h"

#define UNUSED(X) (void) X

void vMyFunc(void)
{
  vTraceBegin();
  
  vTrace(INFO_LEVEL, "INFO MESSAGE");
  vTrace(WARNING_LEVEL, "WARNING MESSAGE");
  vTrace(ERROR_LEVEL, "ERROR MESSAGE");
  vTrace(FATAL_LEVEL, "FATAL MESSAGE");
  vTrace(DEBUG_LEVEL, "DEBUG MESSAGE");
  vTrace(TRACE_LEVEL, "TRACE MESSAGE");

  vTraceEnd();
}

int main(int argc, char **argv)
{
  UNUSED(argc);
  UNUSED(argv);
  UNUSED(gbColoredLogLevel);
  UNUSED(gszLogFileName);
  UNUSED(gszConfFileName);
  UNUSED(kszLogLevelColorEnd);
  UNUSED(kszLogLevelColorInit);
  UNUSED(kszLogLevel);

  vSetConfFileName("log.conf");

  vSetLogLevel(iGetLogLevel());

  if(giDebugLevel < 0)
  {
    fprintf(stderr, "Error, giDebugLevel return value: %d!\n", giDebugLevel);

    exit(EXIT_FAILURE);
  }
  
  vSetLogFileName("trace.log");

  vTraceBegin();
  
  vTrace(INFO_LEVEL, "Main Info Message :)");

  vMyFunc();

  vTraceEnd();

  return 0;
}



src/log.c
/* log.c: This file contains the implementation of logger functions
 *
 * Written by Gustavo Bacagine <gustavo.bacagine@protonmail.com>
 *
 * Date: 2023-06-16
 */

#include "log.h"

char gszConfFileName[LOG_FILE_NAME_LENGTH];
char gszLogFileName[LOG_FILE_NAME_LENGTH];
DebugLevel giDebugLevel = 0;
bool gbColoredLogLevel = false;

void vSetConfFileName(const char *kszConfFileName)
{
  if(sizeof(kszConfFileName) > sizeof(gszConfFileName))
  {
    fprintf(stderr, "E: The name of configuraation file is bigger than %d", CONF_FILE_NAME_LENGTH);
    exit(EXIT_FAILURE);
  }
  
  snprintf(gszConfFileName, sizeof(gszConfFileName), "%s", kszConfFileName); 
}

void vSetLogFileName(const char *kszLogFileName)
{
  if(sizeof(kszLogFileName) > sizeof(gszLogFileName))
  {
    fprintf(stderr, "E: The name of log file is bigger than %d", LOG_FILE_NAME_LENGTH);
    exit(EXIT_FAILURE);
  }
  
  snprintf(gszLogFileName, sizeof(gszLogFileName), "%s", kszLogFileName); 
}

void vSetColoredLogLevel(bool bColored)
{
  gbColoredLogLevel = bColored;
}

void vSetLogLevel(DebugLevel usiDebugLevel)
{
  giDebugLevel = usiDebugLevel;
}

void _vTrace(DebugLevel usiDebugLevel,
             const char *kpszModuleName,
             const int kiLine, 
             const char *kpszFmt, ...)
{
  va_list args;
  FILE *fpLogFile = NULL;
  time_t tTime = time(NULL);
  struct tm *stTime = localtime(&tTime);
  char szDbg[2048];

  memset(szDbg , 0, sizeof(szDbg));

  if((fpLogFile = fopen(gszLogFileName, "a")) == NULL)
  {
    fprintf(stderr, "E: Impossible create or open file %s!\n"
                    "%s\n", gszLogFileName, strerror(errno));

    exit(EXIT_FAILURE);
  }
  
  va_start(args, kpszFmt);
  
  snprintf(szDbg, sizeof(szDbg), "%04d/%02d/%02d %02d:%02d:%02d %s:%d", stTime->tm_year + 1900, 
                                                                        stTime->tm_mon + 1, 
                                                                        stTime->tm_mday,
                                                                        stTime->tm_hour,
                                                                        stTime->tm_min,
                                                                        stTime->tm_sec,
                                                                        kpszModuleName,
                                                                        kiLine
  );

  switch(usiDebugLevel)
  {
    case INFO_LEVEL:
      if(gbColoredLogLevel)
      {
        char szLogLevel[22];
        
        memset(szLogLevel, 0, sizeof(szLogLevel));
        
        strcat(szLogLevel, kszLogLevelColorInit[INFO_LEVEL]);
        strcat(szLogLevel, kszLogLevel[INFO_LEVEL]);
        strcat(szLogLevel, kszLogLevelColorEnd[INFO_LEVEL]);
      
        strcat(szDbg, " [");
        strcat(szDbg, szLogLevel);
        strcat(szDbg, "] ");

        strcat(szDbg, kpszFmt);
        strcat(szDbg, "\n");

        vfprintf(fpLogFile, szDbg, args);
        break;
      }
      
      strcat(szDbg, " [");
      strcat(szDbg, kszLogLevel[INFO_LEVEL]);
      strcat(szDbg, "] ");

      strcat(szDbg, kpszFmt);
      strcat(szDbg, "\n");

      vfprintf(fpLogFile, szDbg, args);
      break;
    case WARNING_LEVEL:
      if(gbColoredLogLevel)
      {
        char szLogLevel[25];
        
        memset(szLogLevel, 0, sizeof(szLogLevel));
        
        strcat(szLogLevel, kszLogLevelColorInit[WARNING_LEVEL]);
        strcat(szLogLevel, kszLogLevel[WARNING_LEVEL]);
        strcat(szLogLevel, kszLogLevelColorEnd[WARNING_LEVEL]);
   
        strcat(szDbg, " [");     
        strcat(szDbg, szLogLevel);
        strcat(szDbg, "] ");

        strcat(szDbg, kpszFmt);
        strcat(szDbg, "\n");

        vfprintf(fpLogFile, szDbg, args);
        break;
      }

      strcat(szDbg, " [");
      strcat(szDbg, kszLogLevel[WARNING_LEVEL]);
      strcat(szDbg, "] ");

      strcat(szDbg, kpszFmt);
      strcat(szDbg, "\n");

      vfprintf(fpLogFile, szDbg, args);
      break;
    case ERROR_LEVEL:
      if(gbColoredLogLevel)
      {
        char szLogLevel[23];
        
        memset(szLogLevel, 0, sizeof(szLogLevel));
        
        strcat(szLogLevel, kszLogLevelColorInit[ERROR_LEVEL]);
        strcat(szLogLevel, kszLogLevel[ERROR_LEVEL]);
        strcat(szLogLevel, kszLogLevelColorEnd[ERROR_LEVEL]);

        strcat(szDbg, " [");
        strcat(szDbg, szLogLevel);
        strcat(szDbg, "] ");

        strcat(szDbg, kpszFmt);
        strcat(szDbg, "\n");

        vfprintf(fpLogFile, szDbg, args);
        break;
      }

      strcat(szDbg, " [");
      strcat(szDbg, kszLogLevel[ERROR_LEVEL]);
      strcat(szDbg, "] ");

      strcat(szDbg, kpszFmt);
      strcat(szDbg, "\n");

      vfprintf(fpLogFile, szDbg, args);
      break;
    case FATAL_LEVEL:
      if(gbColoredLogLevel)
      {
        char szLogLevel[23];
        
        memset(szLogLevel, 0, sizeof(szLogLevel));
        
        strcat(szLogLevel, kszLogLevelColorInit[FATAL_LEVEL]);
        strcat(szLogLevel, kszLogLevel[FATAL_LEVEL]);
        strcat(szLogLevel, kszLogLevelColorEnd[FATAL_LEVEL]);

        strcat(szDbg, " [");
        strcat(szDbg, szLogLevel);
        strcat(szDbg, "] ");

        strcat(szDbg, kpszFmt);
        strcat(szDbg, "\n");

        vfprintf(fpLogFile, szDbg, args);
        break;
      }

      strcat(szDbg, " [");
      strcat(szDbg, kszLogLevel[FATAL_LEVEL]);
      strcat(szDbg, "] ");

      strcat(szDbg, kpszFmt);
      strcat(szDbg, "\n");

      vfprintf(fpLogFile, szDbg, args);
      break;
    case DEBUG_LEVEL:
      if(gbColoredLogLevel)
      {
        char szLogLevel[23];
        
        memset(szLogLevel, 0, sizeof(szLogLevel));
        
        strcat(szLogLevel, kszLogLevelColorInit[DEBUG_LEVEL]);
        strcat(szLogLevel, kszLogLevel[DEBUG_LEVEL]);
        strcat(szLogLevel, kszLogLevelColorEnd[DEBUG_LEVEL]);

        strcat(szDbg, " [");
        strcat(szDbg, szLogLevel);
        strcat(szDbg, "] ");

        strcat(szDbg, kpszFmt);
        strcat(szDbg, "\n");

        vfprintf(fpLogFile, szDbg, args);
        break;
      }

      strcat(szDbg, " [");
      strcat(szDbg, kszLogLevel[DEBUG_LEVEL]);
      strcat(szDbg, "] ");

      strcat(szDbg, kpszFmt);
      strcat(szDbg, "\n");

      vfprintf(fpLogFile, szDbg, args);
      break;
    case TRACE_LEVEL:
      if(gbColoredLogLevel)
      {
        char szLogLevel[23];
        
        memset(szLogLevel, 0, sizeof(szLogLevel));
        
        strcat(szLogLevel, kszLogLevelColorInit[TRACE_LEVEL]);
        strcat(szLogLevel, kszLogLevel[TRACE_LEVEL]);
        strcat(szLogLevel, kszLogLevelColorEnd[TRACE_LEVEL]);

        strcat(szDbg, " [");
        strcat(szDbg, szLogLevel);
        strcat(szDbg, "] ");

        strcat(szDbg, kpszFmt);
        strcat(szDbg, "\n");

        vfprintf(fpLogFile, szDbg, args);
        break;
      }

      strcat(szDbg, " [");
      strcat(szDbg, kszLogLevel[TRACE_LEVEL]);
      strcat(szDbg, "] ");
      
      strcat(szDbg, kpszFmt);
      strcat(szDbg, "\n");

      vfprintf(fpLogFile, szDbg, args);
      break;
    default:
      break;
  }

  va_end(args);

  fclose(fpLogFile);
  fpLogFile = NULL;
}

int iGetLogLevel(void)
{
  FILE *fpLogConfFile = NULL;
  int iDebugLevel = -1;
  char szLine[256];
  int iLineLen = sizeof(szLine);

  /* Used to verify if the LOG_LEVEL 
   * variable is found in .conf file */
  bool bFoundLogLevel = false;
  
  char *pToken = NULL;
  int ii = 0;
  char szLevel[2];
  
  memset(szLine, 0, iLineLen);
  memset(szLevel, 0, sizeof(szLevel));

  /* Try open .conf file */
  if((fpLogConfFile = fopen(gszConfFileName, "r")) == NULL)
  {
    fprintf(stderr, "E: %s %s\n", gszConfFileName, strerror(errno));

    return -1;
  }

  while(fgets(szLine, iLineLen, fpLogConfFile) != NULL)
  {
    /* Ignore commented szLines */
    if(strstr(szLine, "#")) continue;
    pToken = strtok(szLine, " = ");
    
    if(strstr(szLine, "LOG_LEVEL"))
    {
      bFoundLogLevel = true;
      
      while(pToken != NULL)
      {
        /* Getting the value after '=' 
         * symbol and converting form 
         * string to char
         */
        if(ii == 1)
        {
          /* Verify if have a value after '=' symbol */
          if(!strcmp(pToken, "") || !strcmp(pToken, " ") || !strcmp(pToken, "\n"))
          {
            fprintf(stderr, "E: LOG_LEVEL is empty in file %s!\n", gszConfFileName);
            
            fclose(fpLogConfFile);
            fpLogConfFile = NULL;
            
            return -2;
          }

          snprintf(szLevel, sizeof(szLevel), "%s", pToken);
          iDebugLevel = atoi(szLevel);
          
          break;
        }
        pToken = strtok(NULL, " = ");

        ii++;
      }

      break;
    }
    else
    {
      continue;
    }
  }

  /* If don't found the LOG_LEVEL 
   * variable in file print a error 
   * message to user */
  if(bFoundLogLevel == false)
  {
    fprintf(stderr, "E: Not found variable LOG_LEVEL in file %s!\n", gszConfFileName);
        
    fclose(fpLogConfFile);
    fpLogConfFile = NULL;

    return -3;
  }
  
  /* If LOG_LEVEL in file is incorret */
  if(iDebugLevel < 0 || iDebugLevel > 5)
  {
    fprintf(stderr, "E: Invalid value of log level in file %s!\n", gszConfFileName);
        
    fclose(fpLogConfFile);
    fpLogConfFile = NULL;

    return -4;
  }

  fclose(fpLogConfFile);
  fpLogConfFile = NULL;

  return iDebugLevel;
}

int iGetColoredLogLevel(void)
{
  FILE *fpLogConfFile = NULL;
  char szLine[256];
  int iLineLen = sizeof(szLine);

  /* Used to verify if the COLORED_LOG_LEVEL 
   * variable is found in .conf file */
  bool bFoundColoredLogLevel = false;
  
  char *pToken = NULL;
  int ii = 0;
  char szBool[6]; /* Receive true or false */
  int iReturnValue = false;

  memset(szLine, 0, iLineLen);
  memset(szBool, 0, sizeof(szBool));

  /* Try open .conf file */
  if((fpLogConfFile = fopen(gszConfFileName, "r")) == NULL)
  {
    fprintf(stderr, "E: %s %s\n", gszConfFileName, strerror(errno));

    return -1;
  }

  while(fgets(szLine, iLineLen, fpLogConfFile) != NULL)
  {
    /* Ignore commented szLines */
    if(strstr(szLine, "#")) continue;
    pToken = strtok(szLine, " = ");
    
    if(strstr(szLine, "COLORED_LOG_LEVEL"))
    {
      bFoundColoredLogLevel = true;
      
      while(pToken != NULL)
      {
        /* Getting the value after '=' 
         * symbol and converting form 
         * string to char
         */
        if(ii == 1)
        {
          /* Verify if have a value after '=' symbol */
          if(!strcmp(pToken, "") || !strcmp(pToken, " ") || !strcmp(pToken, "\n"))
          {
            fprintf(stderr, "E: COLORED_LOG_LEVEL is empty in file %s!\n", gszConfFileName);
            
            fclose(fpLogConfFile);
            fpLogConfFile = NULL;
            
            return -2;
          }

          snprintf(szBool, sizeof(szBool), "%s", pToken);
          szBool[strcspn(szBool, "\n")] = 0;
          
          break;
        }
        pToken = strtok(NULL, " = ");

        ii++;
      }

      break;
    }
    else
    {
      continue;
    }
  }

  /* If don't found the LOG_LEVEL 
   * variable in file print a error 
   * message to user */
  if(bFoundColoredLogLevel == false)
  {
    fprintf(stderr, "E: Not found variable COLORED_LOG_LEVEL in file %s!\n", gszConfFileName);
        
    fclose(fpLogConfFile);
    fpLogConfFile = NULL;

    return -3;
  }
  
  /* If COLORED_LOG_LEVEL in file is incorret */
  if(!strcmp(szBool, "true") || !strcmp(szBool, "TRUE"))
  {
    iReturnValue = 1;
  }
  else if(!strcmp(szBool, "false") || !strcmp(szBool, "FALSE"))
  {
    iReturnValue = 0;
  }
  else
  {
    fprintf(stderr, "E: Invalid value of colored log level in file %s!\n", gszConfFileName);

    iReturnValue = -4;
  }

  fclose(fpLogConfFile);
  fpLogConfFile = NULL;

  return iReturnValue;
}



include/trace/trace.h
/**
 * trace.h: This file contains constants and functions to create 
 * .log files
 *
 * Written by Gustavo Bacagine <gustavo.bacagine@protonmai.com>
 *
 * Date: 2023-06-16
 */

#ifndef _TRACE_H_
#define _TRACE_H_

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <stdarg.h>
#include <time.h>
#include <errno.h>

#define INFO_LEVEL    1 /* Normal messages                                      */
#define WARNING_LEVEL 2 /* Warning alert messages                               */
#define ERROR_LEVEL   3 /* Error messages                                       */
#define FATAL_LEVEL   4 /* Fatal error messages                                 */
#define DEBUG_LEVEL   5 /* Messages with more information about the source code */
#define TRACE_LEVEL   6 /* Show all messages in trace                           */

/**
 * Example of use:
 *
 * if(INFO_DETAILS)
 * {
 *   vTraceInfo("INFO Message");
 * }
 *
 * if(DEBUG_DETAILS)
 * {
 *   vTraceDebug("DEBUG Message");
 * }
 */
#define INFO_DETAILS    giDebugLevel >= INFO_LEVEL   
#define WARNING_DETAILS giDebugLevel >= WARNING_LEVEL
#define ERROR_DETAILS   giDebugLevel >= ERROR_LEVEL
#define FATAL_DETAILS   giDebugLevel >= FATAL_LEVEL
#define DEBUG_DETAILS   giDebugLevel >= DEBUG_LEVEL
#define TRACE_DETAILS   giDebugLevel >= TRACE_LEVEL

/* White color */
#define INFO_INIT_COLOR "\033[1;37m"
#define INFO_END_COLOR  "\033[0m"

/* Cyan color */
#define WARNING_INIT_COLOR "\033[1;35m"
#define WARNING_END_COLOR  "\033[0m"

/* Red color */
#define ERROR_INIT_COLOR "\033[1;31m"
#define ERROR_END_COLOR  "\033[0m"

#define FATAL_INIT_COLOR "\033[1;31m"
#define FATAL_END_COLOR  "\033[0m"

/* Blue color */
#define DEBUG_INIT_COLOR "\033[1;34m"
#define DEBUG_END_COLOR  "\033[0m"

/* Yellow color */
#define TRACE_INIT_COLOR "\033[1;33m"
#define TRACE_END_COLOR  "\033[0m"

#define vTraceInfo(FORMAT, ...)    _vTrace(INFO_LEVEL,    __FILE__, __LINE__, FORMAT, ##__VA_ARGS__)
#define vTraceWarning(FORMAT, ...) _vTrace(WARNING_LEVEL, __FILE__, __LINE__, FORMAT, ##__VA_ARGS__)
#define vTraceError(FORMAT, ...)   _vTrace(ERROR_LEVEL,   __FILE__, __LINE__, FORMAT, ##__VA_ARGS__)
#define vTraceFatal(FORMAT, ...)   _vTrace(FATAL_LEVEL,   __FILE__, __LINE__, FORMAT, ##__VA_ARGS__)
#define vTraceDebug(FORMAT, ...)   _vTrace(DEBUG_LEVEL,   __FILE__, __LINE__, FORMAT, ##__VA_ARGS__)
#define vTraceAll(FORMAT, ...)     _vTrace(TRACE_LEVEL,   __FILE__, __LINE__, FORMAT, ##__VA_ARGS__)

#define vTrace(LOG_LEVEL, FORMAT, ...) if(giDebugLevel >= LOG_LEVEL) _vTrace(LOG_LEVEL, __FILE__, __LINE__, FORMAT, ##__VA_ARGS__)

#define vTraceBegin() if(INFO_DETAILS) _vTrace(INFO_LEVEL, __FILE__, __LINE__, "%s - begin", __func__)
#define vTraceEnd()   if(INFO_DETAILS) _vTrace(INFO_LEVEL, __FILE__, __LINE__, "%s - end", __func__)

static const char *kszLogLevel[] = {
  "INFO", "WARNING", "ERROR", "FATAL", "DEBUG", "TRACE"
};

static const char *kszLogLevelColorInit[] = {
  INFO_INIT_COLOR, WARNING_INIT_COLOR, 
  ERROR_INIT_COLOR, FATAL_INIT_COLOR, 
  DEBUG_INIT_COLOR, TRACE_INIT_COLOR
};

static const char *kszLogLevelColorEnd[] = {
  INFO_END_COLOR, WARNING_END_COLOR, 
  ERROR_END_COLOR, FATAL_END_COLOR, 
  DEBUG_END_COLOR, TRACE_END_COLOR
};

/**
 * Receive the name of .conf file
 */
extern char gszConfFileName[256];

/**
 * Receive the name of .trace file
 */
extern char gszLogFileName[256]; 

/**
 * Receive the level of debug from .conf 
 * file, the default is 1 (INFO_LEVEL)
 */
extern int giDebugLevel;

/**
 * Receive if the trace level tag is colored 
 * or not, the default is without color
 */
extern bool gbColoredLogLevel;

/**
 * Set the name of configure file of 
 * the software
 */
void vSetConfFileName(const char *kszConfFile);

/**
 * Set the name of trace file
 */
void vSetLogFileName(const char *kszLogFileName);

/** 
 * Enable or disable trace level with 
 * colored and bold
 */
void vSetColoredLogLevel(bool bColored);

/**
 * Set the value of giDebugLevel
 */
void vSetLogLevel(const int kiDebugLevel);

/**
 * Get the trace level from file.
 * 
 * return values:
 * 1 - 6 = Log Level
 * -1 if can't open the file.
 * -2 if LOG_LEVEL is empty in file
 * -3 if don't found LOG_LEVEL variable in file
 * -4 if value of LOG_LEVEL is invalid
 */
int iGetLogLevel(void);

/**
 * Get the colored trace level from file.
 * 
 * return values:
 *  0 is fals, no color in trace level
 *  1 is true, colored in trace level
 * -1 if can't open the file.
 * -2 if COLORED_LOG_LEVEL is empty in file
 * -3 if don't found COLORED_LOG_LEVEL variable in file
 * -4 if value of COLORED_LOG_LEVEL is invalid
 */
int iGetColoredLogLevel(void);

/**
 * Print a trace message with the trace level between [].
 */
void _vTrace(const int kiDebugLevel,
             const char *kpszModuleName,
             const int kiLine,
             const char *kpszFmt, ...);

#endif /* _TRACE_H_ */



include/log/log.h
/* log.h: This file contains constants and functions to create 
 * .log files
 *
 * Written by Gustavo Bacagine <gustavo.bacagine@protonmai.com>
 *
 * Date: 2023-06-16
 */

#ifndef _LOG_H_
#define _LOG_H_

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <stdarg.h>
#include <time.h>
#include <errno.h>

#define CONF_FILE_NAME_LENGTH 256
#define LOG_FILE_NAME_LENGTH  256

#define INFO_LEVEL    0 /* Normal messages                                      */
#define WARNING_LEVEL 1 /* Warning alert messages                               */
#define ERROR_LEVEL   2 /* Error messages                                       */
#define FATAL_LEVEL   3 /* Fatal error messages                                 */
#define DEBUG_LEVEL   4 /* Messages with more information about the source code */
#define TRACE_LEVEL   5 /* Show all messages in log                             */

/**
 * Example of use:
 *
 * if(INFO_DETAILS)
 * {
 *   vTraceInfo("INFO Message");
 * }
 *
 * if(DEBUG_DETAILS)
 * {
 *   vTraceDebug("DEBUG Message");
 * }
 */
#define INFO_DETAILS    giDebugLevel >= INFO_LEVEL   
#define WARNING_DETAILS giDebugLevel >= WARNING_LEVEL
#define ERROR_DETAILS   giDebugLevel >= ERROR_LEVEL
#define FATAL_DETAILS   giDebugLevel >= FATAL_LEVEL
#define DEBUG_DETAILS   giDebugLevel >= DEBUG_LEVEL
#define TRACE_DETAILS   giDebugLevel >= TRACE_LEVEL

/* White color */
#define INFO_INIT_COLOR "\033[1;37m"
#define INFO_END_COLOR  "\033[0m"

/* Cyan color */
#define WARNING_INIT_COLOR "\033[1;35m"
#define WARNING_END_COLOR  "\033[0m"

/* Red color */
#define ERROR_INIT_COLOR "\033[1;31m"
#define ERROR_END_COLOR  "\033[0m"

#define FATAL_INIT_COLOR "\033[1;31m"
#define FATAL_END_COLOR  "\033[0m"

/* Blue color */
#define DEBUG_INIT_COLOR "\033[1;34m"
#define DEBUG_END_COLOR  "\033[0m"

/* Yellow color */
#define TRACE_INIT_COLOR "\033[1;33m"
#define TRACE_END_COLOR  "\033[0m"

#define vTraceInfo(FORMAT, ...)    _vTrace(INFO_LEVEL,    __FILE__, __LINE__, FORMAT, ##__VA_ARGS__)
#define vTraceWarning(FORMAT, ...) _vTrace(WARNING_LEVEL, __FILE__, __LINE__, FORMAT, ##__VA_ARGS__)
#define vTraceError(FORMAT, ...)   _vTrace(ERROR_LEVEL,   __FILE__, __LINE__, FORMAT, ##__VA_ARGS__)
#define vTraceFatal(FORMAT, ...)   _vTrace(FATAL_LEVEL,   __FILE__, __LINE__, FORMAT, ##__VA_ARGS__)
#define vTraceDebug(FORMAT, ...)   _vTrace(DEBUG_LEVEL,   __FILE__, __LINE__, FORMAT, ##__VA_ARGS__)
#define vTraceAll(FORMAT, ...)     _vTrace(TRACE_LEVEL,   __FILE__, __LINE__, FORMAT, ##__VA_ARGS__)

#define vTrace(LOG_LEVEL, FORMAT, ...) if(giDebugLevel >= LOG_LEVEL) _vTrace(LOG_LEVEL, __FILE__, __LINE__, FORMAT, ##__VA_ARGS__)

#define vTraceBegin() if(INFO_DETAILS) _vTrace(INFO_LEVEL, __FILE__, __LINE__, "%s - begin", __func__)
#define vTraceEnd()   if(INFO_DETAILS) _vTrace(INFO_LEVEL, __FILE__, __LINE__, "%s - end", __func__)

typedef int DebugLevel;

static const char *kszLogLevel[] = {
  "INFO", "WARNING", "ERROR", "FATAL", "DEBUG", "TRACE"
};

static const char *kszLogLevelColorInit[] = {
  INFO_INIT_COLOR, WARNING_INIT_COLOR, 
  ERROR_INIT_COLOR, FATAL_INIT_COLOR, 
  DEBUG_INIT_COLOR, TRACE_INIT_COLOR
};

static const char *kszLogLevelColorEnd[] = {
  INFO_END_COLOR, WARNING_END_COLOR, 
  ERROR_END_COLOR, FATAL_END_COLOR, 
  DEBUG_END_COLOR, TRACE_END_COLOR
};

/**
 * Receive the name of .conf file
 */
extern char gszConfFileName[LOG_FILE_NAME_LENGTH];

/**
 * Receive the name of .log file
 */
extern char gszLogFileName[LOG_FILE_NAME_LENGTH]; 

/**
 * Receive the level of debug from .conf 
 * file, the default is 0 (INFO_LEVEL)
 */
extern DebugLevel giDebugLevel;

/**
 * Receive if the log level tag is colored 
 * or not, the default is without color
 */
extern bool gbColoredLogLevel;

/**
 * Set the name of configure file of 
 * the software
 */
void vSetConfFileName(const char *kszConfFile);

/**
 * Set the name of log file
 */
void vSetLogFileName(const char *kszLogFileName);

/** 
 * Enable or disable log level with 
 * colored and bold
 */
void vSetColoredLogLevel(bool bColored);

/**
 * Set the value of giDebugLevel
 */
void vSetLogLevel(DebugLevel usiDebugLevel);

/**
 * Get the log level from file.
 * 
 * return values:
 * 0 - 6 = Log Level
 * -1 if can't open the file.
 * -2 if LOG_LEVEL is empty in file
 * -3 if don't found LOG_LEVEL variable in file
 * -4 if value of LOG_LEVEL is invalid
 */
int iGetLogLevel(void);

/**
 * Get the colored log level from file.
 * 
 * return values:
 *  0 is fals, no color in log level
 *  1 is true, colored in log level
 * -1 if can't open the file.
 * -2 if COLORED_LOG_LEVEL is empty in file
 * -3 if don't found COLORED_LOG_LEVEL variable in file
 * -4 if value of COLORED_LOG_LEVEL is invalid
 */
int iGetColoredLogLevel(void);

/**
 * Print a log message with the log level between [].
 */
void _vTrace(const DebugLevel usiDebugLevel,
             const char *kpszModuleName,
             const int kiLine,
             const char *kpszFmt, ...);

#endif /* _LOG_H_ */

